### PR TITLE
Fix Default Asset File locator for clang docs

### LINKED ISSUE #98056: Restore functionality to clang-doc asset locator in multi-config builds
Multi-config builds fail to build the clang-doc project, because two asset files aren't found, when the executable running the test runs under the Debug config.
This is because the asset locator expects that the executable running the test is located at the top level build directory, not under an extra directory (e.g., Debug).
AC: fix the asset locator so that the project builds / passes in any configuration.

### PR DESCRIPTION
**Pull request by bob80905**
In `clang-tools-extra\clang-doc\tool\ClangDocMain.cpp`, there is a function, `getDefaultAssetFiles` that tries to find the location of 
`<build dir>\share\clang-doc\index.js`, but fails. This is because there are alternate configurations in setting up the llvm project, and executables in different locations may be run. In some cases, executables are built and run in this location:
`<build dir>\bin\<executables>`
and in other cases, this location:
`<build dir>\[Debug | Release]\bin\<executables>`

The locator assumed that the executable that is running was in the first example above, and only ascended one directory, and in the case that an executable under Debug / Release was run, the index.js file wouldn't be found.
This PR changes the way in which the asset files are copied over to the build output directory, by placing the directory containing the asset files directly under whichever specific build-output configuration directory is being processed by cmake.
In this way, regardless of which configuration's executable runs `getDefaultAssetFiles`, the asset files will be found directly underneath the current build directory.
Fixes https://github.com/llvm/llvm-project/issues/98056

### CODE REVIEW

**Comment by bob80905** [comment]
An important assumption I'm making here is that the Debug directory cannot exist without the Release directory. If this assumption doesn't hold, then this fix will fail if only the release directory exists and debug is missing. I believe by the nature of how the llvm project is generated, both directories are generated at all times.

**Review by bogner — CHANGES_REQUESTED** [review]
I don't think this is the right fix. The Release/Debug directories only exist in the build tree, so this is incurring an unnecessary cost in an installed clang-doc and introducing a potentially very strange behaviour for weird install locations.

Probably the simplest approach to fixing the problem would be to duplicate the share/clang-doc directory per config. This is a bit annoying because of the redundancy, but it's probably better than pursuing something more elaborate.

File: clang-tools-extra/clang-doc/tool/ClangDocMain.cpp [hunk 1]
```diff
@@ -167,7 +167,14 @@ llvm::Error getDefaultAssetFiles(const char *Argv0,
 
   llvm::SmallString<128> AssetsPath;
   AssetsPath = llvm::sys::path::parent_path(NativeClangDocPath);
-  llvm::sys::path::append(AssetsPath, "..", "share", "clang-doc");
+  llvm::sys::path::append(AssetsPath, "..");
+  llvm::SmallString<128> tempCopyDbg;
+  llvm::sys::path::native(AssetsPath, tempCopyDbg);
+  llvm::sys::path::append(tempCopyDbg, "Debug");
```
Comment (llvm-beanz) [inline clang-tools-extra/clang-doc/tool/ClangDocMain.cpp:173]: Could this also be "Release"?

It seems like this should probably be built like the compiler resource directory is where the assets get installed into a known binary relative path.

**Comment by damyanp** [comment]
I don't think that the current description matches the change?

Does this mean that there'll likely be 3 copies of each file.  One per configuration, and another copy just in the binary directory?

**Comment by bob80905** [comment]
Updated the description.
In my tests, there are at least 5 copies of this file in the overall build output directory:
```
<llvm-project-dir>/<MS-Build-Dir>/Debug/share/clang-doc
<llvm-project-dir>/<MS-Build-Dir>/Release/share/clang-doc
<llvm-project-dir>/<MS-Build-Dir>/MinSizeRel/share/clang-doc
<llvm-project-dir>/<MS-Build-Dir>/RelWithDebInfo/share/clang-doc
<llvm-project-dir>/<MS-Build-Dir>/share/clang-doc
```
Yes, one per configuration, and one underneath the binary directory.
I am unsure if we still need the copy in the binary directory, because I'm not sure which executable is run by default when running tests.

**Comment by bogner** [comment]
> Updated the description. In my tests, there are at least 5 copies of this file in the overall build output directory:
> 
> ```
> <llvm-project-dir>/<MS-Build-Dir>/Debug/share/clang-doc
> <llvm-project-dir>/<MS-Build-Dir>/Release/share/clang-doc
> <llvm-project-dir>/<MS-Build-Dir>/MinSizeRel/share/clang-doc
> <llvm-project-dir>/<MS-Build-Dir>/RelWithDebInfo/share/clang-doc
> <llvm-project-dir>/<MS-Build-Dir>/share/clang-doc
> ```
> 
> Yes, one per configuration, and one underneath the binary directory. I am unsure if we still need the copy in the binary directory, because I'm not sure which executable is run by default when running tests.

We shouldn't need the one in the binary directory - IIUC one per configuration will be sufficient, and the one that's never used is potentially confusing.

Please also test locally that the install target does the right thing with these changes. You can use `-DCMAKE_INSTALL_PREFIX=/some/temp/dir` to make sure it doesn't install to some system location and is easy to spot check.

File: clang-tools-extra/clang-doc/tool/CMakeLists.txt [hunk 2]
```diff
@@ -44,7 +42,10 @@ foreach(f ${assets})
   install(FILES ${asset_dir}/${f}
     DESTINATION "${CMAKE_INSTALL_DATADIR}/clang-doc"
     COMPONENT clang-doc)
-  copy_files_to_dst(${asset_dir} ${resource_dir} ${f})
+   foreach(config ${CMAKE_CONFIGURATION_TYPES})
```
[the remark this answers is not in the document]
Comment (bob80905) [reply clang-tools-extra/clang-doc/tool/CMakeLists.txt:45]: No, ninja / other single-config build systems fail, and such configs expect the asset files to be under 
`<llvm-project-dir>/<MS-Build-Dir>/share/clang-doc`

File: clang-tools-extra/clang-doc/tool/CMakeLists.txt [hunk 3]
```diff
@@ -25,17 +25,15 @@ set(assets
...
-  add_custom_command(OUTPUT ${dst}
-    DEPENDS ${src}
-    COMMAND ${CMAKE_COMMAND} -E copy_if_different ${src} ${dst}
-    COMMENT "Copying ${file} to ${dst_dir}"
-  )
+  if (NOT EXISTS ${dst_dir})
+    file(MAKE_DIRECTORY ${dst_dir})
+  endif()
+  configure_file(${src} ${dst} COPYONLY)
   list(APPEND out_files ${dst})
```
Comment (damyanp) [inline clang-tools-extra/clang-doc/tool/CMakeLists.txt:37]: Why the switch from `add_custom_command` to make directory / `configure_file`?  Is that a required part of this change?  Might be worth mentioning in the commit description.

**Review by damyanp — APPROVED** [review]

File: clang-tools-extra/clang-doc/tool/CMakeLists.txt (hunk 3 above)
Comment (bob80905) [reply clang-tools-extra/clang-doc/tool/CMakeLists.txt]: Unnecessary, probably a remnant of some testing / experimentation

File: clang-tools-extra/clang-doc/tool/CMakeLists.txt [hunk 4]
```diff
@@ -25,12 +25,14 @@ set(assets
...
 set(asset_dir "${CMAKE_CURRENT_SOURCE_DIR}/../assets")
-set(resource_dir "${CMAKE_BINARY_DIR}/share/clang-doc")
 set(out_files)
 
 function(copy_files_to_dst src_dir dst_dir file)
   set(src "${src_dir}/${file}")
   set(dst "${dst_dir}/${file}")
+  if (NOT EXISTS ${dst_dir})
+    file(MAKE_DIRECTORY ${dst_dir})
+  endif()
```
Comment (bogner) [inline clang-tools-extra/clang-doc/tool/CMakeLists.txt:35]: I don't believe this is necessary - copy_if_different will do the right thing.

**Review by bogner — APPROVED** [review]
It's kind of unfortunate how this works dependency-wise - we end up creating "Debug/share/clang-doc/${f}" even when we only build "Release" and vice-versa, as the "copy-clang-doc-assets" explicitly depends on each version of this file. Also the actual install copies the file directly from the source directory, so if we did end up actually configuring these files per config for some reason there's some potential for confusion.

However, the cmake-fu to do this any better is beyond me, and the current state is that we actually have broken bots because of this, so I think the solution here is Good Enough to get in and get the bots unblocked.

**Comment by bob80905** [comment]
Closing this PR since https://github.com/llvm/llvm-project/pull/98099 resolves the issue and has already been merged.

**Pull request closed without merging**